pkg/sideinputs/utils/utils.go

```diff
@@ -0,0 +1,48 @@
+package utils
+
+import (
+	"fmt"
+	"os"
+)
+
+// CheckFileExists checks if a file with the given fileName exists in the file system.
+func CheckFileExists(fileName string) bool {
+	_, err := os.Stat(fileName)
+	// check if err is "file not exists"
+	return !os.IsNotExist(err)
+}
+
+// UpdateSideInputFile writes the given value to the Side input file specified.
+func UpdateSideInputFile(filePath string, value []byte) error {
+	// If the file does not exist, create a new file
+	if !CheckFileExists(filePath) {
+		f, err := os.Create(filePath)
+		if err != nil {
+			return fmt.Errorf("failed to create Side Input file%s : %w", filePath, err)
+		}
+		err = f.Close()
+		if err != nil {
+			return err
+		}
+	}
+	f, err := os.Create(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to create Side input file: %w", err)
+	}
+	defer f.Close()
+	_, err = f.Write(value)
```

Comment:
this is not atomic, if the read happens after Create and before the completion of Write, the user will read the incomplete file.

please use symlinking and renaming to achieve the same

Reply:
Exploring fsnotify and symlinking to check what could be an efficient way for this in our use case.

Reply:
fsnotify?

Reply:
https://pkg.go.dev/github.com/fsnotify/fsnotify@v1.6.0
Which uses the underlying OS signaling system

Reply:
By `renaming` you meant create a new file then replace the old file with the new file?

Reply:
We could write to a new file, update the symlink and then delete the old file

Reply:
> https://pkg.go.dev/github.com/fsnotify/fsnotify@v1.6.0 Which uses the underlying OS signaling system

I am still wondering how fsnotify can solve this. 🤔

Reply:
We would have an event when the symlink or the file is created/written to. Can watch on those on the reader side

Reply:
I'm checking when exactly are these events emitted, and then accordingly see what would work for us.

Reply:
that is on the reader's side, not on the writer's; this code is on the writer's side.

Reply:
> By `renaming` you meant create a new file then replace the old file with the new file?

```
/tmp@vigith$ touch v1 v2           
/tmp@vigith$ ln -s v1 current
/tmp@vigith$ ln -s v2 stage
/tmp@vigith$ ls -l stage current 
lrwxr-xr-x  1 vigith  wheel  2 Aug  8 13:11 current -> v1
lrwxr-xr-x  1 vigith  wheel  2 Aug  8 13:11 stage -> v2
/tmp@vigith$ mv -h stage current
/tmp@vigith$ ls -l current
lrwxr-xr-x  1 vigith  wheel  2 Aug  8 13:11 current -> v2
/tmp@vigith$ 
```

`mv` uses `rename` syscall

Reply:
so in the next iteration you can do something like

```
/tmp@vigith$ rm v1
/tmp@vigith$ touch v3   
/tmp@vigith$ ln -s v3 stage
/tmp@vigith$ ls -l current stage
lrwxr-xr-x  1 vigith  wheel  2 Aug  8 13:11 current -> v2
lrwxr-xr-x  1 vigith  wheel  2 Aug  8 13:23 stage -> v3
/tmp@vigith$ mv stage current
/tmp@vigith$ ls -l current      
lrwxr-xr-x  1 vigith  wheel  2 Aug  8 13:23 current -> v3
/tmp@vigith$ 
```